sessionctx/variable/sysvar.go

```diff
@@ -107,7 +107,7 @@ type SysVar struct {
 func (sv *SysVar) ValidateFromType(vars *SessionVars, value string, scope ScopeFlag) (string, error) {
 	// Some sysvars are read-only. Attempting to set should always fail.
 	if sv.ReadOnly || sv.Scope == ScopeNone {
-		return value, ErrReadOnly.GenWithStackByArgs(sv.Name)
+		return value, errors.Errorf("Variable '%s' is a read only variable", sv.Name)
```

Comment:
The errors.Errorf will change the error code to the "unspecified" value. I tested this in MySQL, and the expected error code is 1238:

```sql
mysql [localhost:8022] {msandbox} (test) > set global version_comment = '1234';
ERROR 1238 (HY000): Variable 'version_comment' is a read only variable
```

I believe this should return the correct code:
```suggestion
		return value, ErrIncorrectScope.GenWithStackByArgs(sv.Name, "read only")
```
(please test it)

Reply:
I have tested it and the error should be `ErrIncorrectScope` and fixed the tests.